### ISSUE TITLE
NetCDF: metadata dictionary contains entries for different variables, also `_FillValue` is confusing

### ISSUE THREAD

**Issue by lnicola** [OSGeo/gdal#6367]
## Expected behavior and actual behavior.

GDAL seems to return metadata entries from a different variable. It also behaves in a confusing manner, but that's probably because `_FillValue` is weird and has two meanings. Apologies if this is normal, I'm not really about the subtleties here.

## Steps to reproduce the problem.

See attached dataset. It has a lot of variables, but those of interest are `oswLandFlag` and `oswPartitions`.

```
$ gdalinfo NETCDF:s1a-iw-ocn-vv-20220412t054445-20220412t054513-042736-051989-001.nc:oswLandFlag
Driver: netCDF/Network Common Data Format
Files: s1a-iw-ocn-vv-20220412t054445-20220412t054513-042736-051989-001.nc
Size is 1, 1
Metadata:
  NC_GLOBAL#bathySource=GEBCO
  NC_GLOBAL#clmSource=GSHHS
  NC_GLOBAL#firstMeasurementTime=2022-04-12T05:44:45.000000Z
  NC_GLOBAL#gmf=CMOD5n
  NC_GLOBAL#iceSource=EUMETSAT OSI SAF
  NC_GLOBAL#idlVersion=
  NC_GLOBAL#IPFversion=003.51
  NC_GLOBAL#lastMeasurementTime=2022-04-12T05:45:13.000000Z
  NC_GLOBAL#missionName=S1A
  NC_GLOBAL#missionPhase=Operational
  NC_GLOBAL#oswAlgorithmVersion=S-1 OSW rev 1/4
  NC_GLOBAL#oswNoiseCorrection=
  NC_GLOBAL#owiAlgorithmVersion=S-1 OWI rev 2/1
  NC_GLOBAL#owiNoiseCorrection=True
  NC_GLOBAL#polarisation=VV
  NC_GLOBAL#polarisationRatio=Identity
  NC_GLOBAL#prf=486.4863102995529
  NC_GLOBAL#processingCenter=MTI
  NC_GLOBAL#processingStartTime=2022-04-12T06:44:52.985266Z
  NC_GLOBAL#rvlAlgorithmVersion=S-1 RVL rev 1/5
  NC_GLOBAL#sourceProduct=S1A_IW_SL2__1_DV_20220412T054446_20220412T054516_042736_051989_BB0C.SAFE
  NC_GLOBAL#statevectorAcc={0,0,0}
  NC_GLOBAL#statevectorPos={5207105.969186,1042893.00983,4667332.776749}
  NC_GLOBAL#statevectorUtc=12-APR-2022 05:45:00.925420
  NC_GLOBAL#statevectorVel={5135.749944,-817.545417,-5531.931286}
  NC_GLOBAL#stokesSource=MODEL WAVEWATCH III (R) GLOBMULTI
  NC_GLOBAL#title=Sentinel-1 OCN Product
  NC_GLOBAL#windSource=ECMWF
  oswLandFlag#flag_meanings=no_land land
  oswLandFlag#flag_values={0,1}
  oswLandFlag#long_name=land flag : 1 if presence of more than 10% of land coverage, 0 instead
  oswLandFlag#valid_range={0,1}
  oswPartitions#long_name=wave partition numbers
  oswPartitions#units=1
  oswPartitions#_FillValue=-128
Corner Coordinates:
Upper Left  (    0.0,    0.0)
Lower Left  (    0.0,    1.0)
Upper Right (    1.0,    0.0)
Lower Right (    1.0,    1.0)
Center      (    0.5,    0.5)
Band 1 Block=1x1 Type=Byte, ColorInterp=Undefined
  Metadata:
    flag_meanings=no_land land
    flag_values={0,1}
    long_name=land flag : 1 if presence of more than 10% of land coverage, 0 instead
    NETCDF_VARNAME=oswLandFlag
    valid_range={0,1}
  Image Structure Metadata:
    PIXELTYPE=SIGNEDBYTE
```

Notice how:

 - part of the metadata is for `oswPartitions`
 - no `_FillValue` is reported for `oswLandFlag`

Using `netCDF4`:

```
In [14]: ds["oswPartitions"]
Out[14]: 
<class 'netCDF4._netCDF4.Variable'>
int8 oswPartitions(oswAzSize, oswRaSize, oswAngularBinSize, oswWavenumberBinSize)
    _FillValue: -128
    units: 1
    long_name: wave partition numbers
unlimited dimensions: 
current shape = (1, 1, 1, 1)
filling on

In [15]: ds["oswLandFlag"]
Out[15]: 
<class 'netCDF4._netCDF4.Variable'>
int8 oswLandFlag(oswAzSize, oswRaSize)
    long_name: land flag : 1 if presence of more than 10% of land coverage, 0 instead
    valid_range: [0 1]
    flag_values: [0 1]
    flag_meanings: no_land land
unlimited dimensions: 
current shape = (1, 1)
filling on, default _FillValue of -127 ignored
```

Not shown, but `GetNoDataValue` returns `None` for `oswLandFlag` and `-128` for `oswPartitions`, which is probably expected given the above. Just like `gdalinfo`, `GetMetadata` on `oswLandFlag` includes the `oswPartitions` entries.

## Operating system

Arch Linux

## GDAL version and provenance

GDAL 3.5.1, distro package

[s1a-iw-ocn-vv-20220412t054445-20220412t054513-042736-051989-001.zip](https://github.com/OSGeo/gdal/files/9575576/s1a-iw-ocn-vv-20220412t054445-20220412t054513-042736-051989-001.zip)

**Comment by rouault** [comment]
The issue about reporting metadata for oswPartitions when opening the oswLandFlag variable is dealt with https://github.com/OSGeo/gdal/pull/6372

Regarding the default fill value of -127 for oswLandFlag not being reported by GDAL as nodata, we follow the recommandation of https://docs.unidata.ucar.edu/netcdf-c/current/attribute_conventions.html : "However, use of the default fill value for data type byte is not recommended."
So basically data producers should specify a _FillValue for variables of type byte is they want to avoid any ambiguity.
